### PR TITLE
feat(kuma-cp): always add kuma.io/zone label to resource

### PR DESCRIPTION
**Pull request by Automaat**
This change is needed to simplify targeting new MeshService in policies. It will simplify how we build subset from MeshService: https://github.com/kumahq/kuma/pull/10422/files#diff-aa3f3b4f30ce7a99336aefc7674797950e7e9c608e6a1c04699452820a8c5041R216

### Checklist prior to review

- [x] [Link to relevant issue][1] as well as docs and UI issues --
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

**Comment by lobkovilya** [comment]
We also need a validation (and unfortunately I missed that for `kuma.io/policy-role` as well), it shouldn't be possible for user to set arbitrary values for `kuma.io/zone` and `kuma.io/policy-role`

**Comment by Automaat** [comment]
> We also need a validation (and unfortunately I missed that for `kuma.io/policy-role` as well), it shouldn't be possible for user to set arbitrary values for `kuma.io/zone` and `kuma.io/policy-role`

I've added validation

**Review by slonka — APPROVED** [review]
LGTM

**Comment by Automaat** [comment]
I had to remove k8s validation. On Kubernetes mutating webhook runs before validating, so we were adding labels and then checking if they are absent

**Review by lobkovilya — CHANGES_REQUESTED** [review]
Please check how we implemented `kuma.io/origin` validation. Because mutating webhook runs before the validating we have to call validation inside the mutating webhook https://github.com/kumahq/kuma/blob/d89e3e53960da72f261229f699b22c3e2d3273e8/pkg/plugins/runtime/k8s/webhooks/defaulter.go#L70

**Comment by Automaat** [comment]
@lobkovilya I've added validation as you suggested

File: pkg/api-server/resource_endpoints.go [hunk 1]
```diff
@@ -532,6 +532,13 @@ func (r *resourceEndpoints) validateLabels(rm model.ResourceMeta) validators.Val
 		}
 	}
 
+	if _, ok := rm.GetLabels()[mesh_proto.PolicyRoleLabel]; ok {
+		err.AddViolationAt(validators.Root().Key(mesh_proto.PolicyRoleLabel), fmt.Sprintf("%s label should not be set manually", mesh_proto.PolicyRoleLabel))
```
Comment (lobkovilya) [inline pkg/api-server/resource_endpoints.go:536]: should be the same as on k8s, we allow these labels to be set, but validate the value to be correct

Comment (Automaat) [reply pkg/api-server/resource_endpoints.go]: right, I've forgot to update it after fixing k8s validation

File: pkg/api-server/resource_endpoints.go [hunk 2]
```diff
@@ -532,17 +532,54 @@ func (r *resourceEndpoints) validateLabels(rm model.ResourceMeta) validators.Val
...
 
-	for _, k := range maps.SortedKeys(rm.GetLabels()) {
+	if r.mode != config_core.Global {
+		zoneTag, ok := resource.GetMeta().GetLabels()[mesh_proto.ZoneTag]
+		if ok && zoneTag != r.zoneName {
+			err.AddViolationAt(validators.Root().Key(mesh_proto.ZoneTag), fmt.Sprintf("%s label should have %s value", mesh_proto.ZoneTag, r.zoneName))
+		}
+	}
+
+	r.validatePolicyRole(resource)
```
Comment (lobkovilya) [inline pkg/api-server/resource_endpoints.go:542]: please validate `r.descriptor` to have `.IsPolicy` and `.IsPluginOriginated`, you can avoid handling cast errors in this case

File: pkg/api-server/resource_endpoints.go [hunk 3]
```diff
@@ -532,17 +532,54 @@ func (r *resourceEndpoints) validateLabels(rm model.ResourceMeta) validators.Val
...
+		for _, msg := range validation.IsValidLabelValue(resource.GetMeta().GetLabels()[k]) {
 			err.AddViolationAt(validators.Root().Key(k), msg)
 		}
 	}
 	return err
 }
 
+func (r *resourceEndpoints) validatePolicyRole(resource rest.Resource) validators.ValidationError {
+	var err validators.ValidationError
+	ns, ok := resource.GetMeta().GetLabels()[mesh_proto.KubeNamespaceTag]
```
Comment (lobkovilya) [inline pkg/api-server/resource_endpoints.go:557]: the code runs only on Universal no? why do we expect `k8s.kuma.io/namespace`?

File: pkg/plugins/runtime/k8s/webhooks/resourceadmissionchecker.go [hunk 4]
```diff
@@ -78,6 +83,42 @@ func (c *ResourceAdmissionChecker) isPrivilegedUser(allowedUsers []string, userI
...
+	return c.validatePolicyRole(r)
+}
+
+func (c *ResourceAdmissionChecker) validatePolicyRole(r core_model.Resource) *admission.Response {
+	ns, ok := r.GetMeta().GetLabels()[mesh_proto.KubeNamespaceTag]
+	if !ok {
+		return nil
+	}
+	policy, ok := r.GetSpec().(core_model.Policy)
+	if !ok {
```
Comment (lobkovilya) [inline pkg/plugins/runtime/k8s/webhooks/resourceadmissionchecker.go:103]: we can check r.Descriptor somewhere earlier in the code, to avoid handling cast errors

File: pkg/plugins/runtime/k8s/webhooks/resourceadmissionchecker.go [hunk 5]
```diff
@@ -41,6 +42,10 @@ func (c *ResourceAdmissionChecker) IsOperationAllowed(userInfo authenticationv1.
 		return c.resourceIsNotAllowedResponse()
 	}
 
+	if errResponse := c.validateLabels(r); errResponse != nil {
```
Comment (lobkovilya) [inline pkg/plugins/runtime/k8s/webhooks/resourceadmissionchecker.go:45]: The method `IsOperationAllowed` calls `isResourceTypeAllowed` to check if "TypeMeta" (what we call Descriptor) is allowed in the environment, then it calls `isResourceAllowed` to check that ResourceMeta+ResourceSpec are allowed in the environment. Your call to `validateLabels` should be moved inside `isResourceAllowed` as it checks ResourceMeta

File: pkg/plugins/runtime/k8s/webhooks/testdata/validation/cli-user_create_dpp.input.yaml [hunk 6]
```diff
@@ -8,6 +8,8 @@ metadata:
   creationTimestamp: null
   annotations:
     k8s.kuma.io/synced: "true"
+  labels:
```
Comment (lobkovilya) [inline pkg/plugins/runtime/k8s/webhooks/testdata/validation/cli-user_create_dpp.input.yaml:11]: why did you add `kuma.io/zone` label here? does it work without the zone label?

Comment (Automaat) [reply pkg/plugins/runtime/k8s/webhooks/testdata/validation/cli-user_create_dpp.input.yaml:11]: it does not work, without zone label now

File: pkg/api-server/resource_endpoints.go (hunk 2 above)
Comment (Automaat) [reply pkg/api-server/resource_endpoints.go]: added

File: pkg/api-server/resource_endpoints.go (hunk 3 above)
Comment (Automaat) [reply pkg/api-server/resource_endpoints.go]: right, removed

File: pkg/plugins/runtime/k8s/webhooks/resourceadmissionchecker.go (hunk 5 above)
Comment (Automaat) [reply pkg/plugins/runtime/k8s/webhooks/resourceadmissionchecker.go]: moved

File: pkg/plugins/runtime/k8s/webhooks/resourceadmissionchecker.go (hunk 4 above)
Comment (Automaat) [reply pkg/plugins/runtime/k8s/webhooks/resourceadmissionchecker.go:100]: changed

File: pkg/plugins/runtime/k8s/webhooks/testdata/validation/cli-user_create_dpp.input.yaml (hunk 6 above)
Comment (lobkovilya) [reply pkg/plugins/runtime/k8s/webhooks/testdata/validation/cli-user_create_dpp.input.yaml:11]: why? it's a breaking change then no? Can't we just add the `kuma.io/zone` label if it's missing?

File: pkg/api-server/resource_endpoints.go [hunk 7]
```diff
@@ -532,17 +532,45 @@ func (r *resourceEndpoints) validateLabels(rm model.ResourceMeta) validators.Val
...
 		}
 	}
 	return err
 }
 
+func (r *resourceEndpoints) validatePolicyRole(resource rest.Resource) validators.ValidationError {
+	var err validators.ValidationError
+	policyObject, ok := resource.GetSpec().(core_model.Policy)
+	if !ok {
+		return err
```
Comment (lobkovilya) [inline pkg/api-server/resource_endpoints.go:561]: you're returning an empty error here, I think it's easier to not handle cast because you're checking the descriptor

File: pkg/api-server/resource_endpoints.go [hunk 8]
```diff
@@ -532,17 +532,45 @@ func (r *resourceEndpoints) validateLabels(rm model.ResourceMeta) validators.Val
...
+	}
+	policyRole, ok := resource.GetMeta().GetLabels()[mesh_proto.PolicyRoleLabel]
+	if !ok {
+		return err
```
Comment (lobkovilya) [inline pkg/api-server/resource_endpoints.go:565]: same, you're returning an empty error, just use [PolicyRole](https://github.com/kumahq/kuma/blob/d89e3e53960da72f261229f699b22c3e2d3273e8/pkg/core/resources/model/resource.go#L493) it always returns a value

Comment (Automaat) [reply pkg/api-server/resource_endpoints.go:565]: I am returning empty error since missing policy role is not a problem here. I cannot use `PolicyRole` method here, as this the resource types are not compatible. This is a `rest.Resource`

File: pkg/api-server/resource_endpoints.go (hunk 7 above)
Comment (Automaat) [reply pkg/api-server/resource_endpoints.go]: changed

File: pkg/plugins/runtime/k8s/webhooks/testdata/validation/cli-user_create_dpp.input.yaml (hunk 6 above)
Comment (Automaat) [reply pkg/plugins/runtime/k8s/webhooks/testdata/validation/cli-user_create_dpp.input.yaml]: I've check again and it is not needed, I don't remember why I've added it

File: pkg/api-server/resource_endpoints.go (hunk 8 above)
Comment (lobkovilya) [reply pkg/api-server/resource_endpoints.go:562]: Just replace `PolicyRole` method with:
```go
func PolicyRole(rm ResourceMeta) mesh_proto.PolicyRole {
	if rm.GetLabels() == nil || rm.GetLabels()[mesh_proto.PolicyRoleLabel] == "" {
		return mesh_proto.SystemPolicyRole
	}
	return mesh_proto.PolicyRole(rm.GetLabels()[mesh_proto.PolicyRoleLabel])
}
```
it doesn't actually need the whole Resource

Comment (Automaat) [reply pkg/api-server/resource_endpoints.go:562]: changed

File: pkg/plugins/runtime/k8s/webhooks/testdata/validation/cli-user_create_zone-ns-mt.input.yaml [hunk 9]
```diff
@@ -6,6 +6,8 @@ metadata:
   labels:
     kuma.io/origin: zone
     kuma.io/mesh: default
+    kuma.io/zone: zone-1
```
Comment (lobkovilya) [inline pkg/plugins/runtime/k8s/webhooks/testdata/validation/cli-user_create_zone-ns-mt.input.yaml:9]: Should we remove these in a couple of places? Maybe only in old test cases. I think it's important to check `kuma.io/zone` label is still optional

**Review by lobkovilya — APPROVED** [review]
LGTM, just a small comment regarding tests

File: pkg/plugins/runtime/k8s/webhooks/testdata/validation/cli-user_create_zone-ns-mt.input.yaml (hunk 9 above)
Comment (Automaat) [reply pkg/plugins/runtime/k8s/webhooks/testdata/validation/cli-user_create_zone-ns-mt.input.yaml]: done
